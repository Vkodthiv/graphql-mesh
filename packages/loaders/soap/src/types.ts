export interface WSDLDefinitionAttributes {
  name: string;
  targetNamespace: string;
}

export interface WSDLDefinition {
  attributes: WSDLDefinitionAttributes;
  'wsdl:import'?: WSDLImport[];
  'wsdl:service'?: WSDLService[];
  'wsdl:binding'?: WSDLBinding[];
  'wsdl:types'?: WSDLTypes[];
  'wsdl:message'?: WSDLMessage[];
  'wsdl:portType'?: WSDLPortType[];
}

export interface WSDLPortTypeAttributes {
  name: string;
}

export interface WSDLPortType {
  attributes: WSDLPortTypeAttributes;
  'wsdl:operation': WSDLOperation[];
}

export interface WSDLMessageAttributes {
  name: string;
}

export interface WSDLPartAttributes {
  element: string;
  name: string;
}

export interface WSDLPart {
  attributes: WSDLPartAttributes;
}

export interface WSDLMessage {
  attributes: WSDLMessageAttributes;
  'wsdl:part': WSDLPart[];
}

export interface XSSchemaAttributes {
  targetNamespace: string;
  version: string;
}

export interface XSSchema {
  attributes: XSSchemaAttributes;
  'xs:complexType': XSComplexType[];
}

export interface XSComplexTypeAttributes {
  name: string;
}

export interface XSElementAttributes {
  name: string;
  type: string;
}

export interface XSElement {
  attributes: XSElementAttributes;
}

export interface XSSequence {
  'xs:element': XSElement[];
}

export interface XSExtensionAttributes {
  base: string;
}

export interface XSExtension {
  attributes: XSExtensionAttributes;
  'xs:sequence': XSSequence[];
}

export interface XSComplexContent {
  'xs:extension': XSExtension[];
}

export interface XSComplexType {
  attributes: XSComplexTypeAttributes;
  'xs:complexContent': XSComplexContent[];
  'xs:sequence': XSSequence[];
}

export interface WSDLTypes {
  'xs:schema': XSSchema[];
}

export interface WSDLBindingAttributes {
  name: string;
  type: string;
}

export interface SOAPBindingAttributes {
  style: 'document';
  transport: string;
}

export interface SOAPBinding {
  attributes: SOAPBindingAttributes;
}

export interface WSDLBinding {
  attributes: WSDLBindingAttributes;
  'soap12:binding': SOAPBinding;
  'wsdl:operation': WSDLOperation[];
}

export interface WSDLOperationAttributes {
  name: string;
}

export interface SOAPOperationAttributes {
  soapAction: string;
  style: string;
}

export interface SOAPOperation {
  attributes: SOAPOperationAttributes;
}

export interface WSDLInputAttributes {
  name: string;
  message: string;
}

export interface SOAPBodyAttributes {
  use: string;
}

export interface SOAPBody {
  attributes: SOAPBodyAttributes;
}

export interface WSDLInput {
  attributes: WSDLInputAttributes;
  'soap12:body': SOAPBody[];
}

export interface WSDLOutputAttributes {
  name: string;
  message: string;
}

export interface WSDLOutput {
  attributes: WSDLOutputAttributes;
  'soap12:body': SOAPBody[];
}

export interface WSDLFaultAttributes {
  name: string;
}

export interface SOAPFault {
  attributes: SOAPFaultAttributes;
}

export interface SOAPFaultAttributes {
  name: string;
  use: string;
}

export interface WSDLFault {
  attributes: WSDLFaultAttributes;
  'soap12:fault': SOAPFault[];
}

export interface WSDLOperation {
  attributes: WSDLOperationAttributes;
  'soap12:operation'?: SOAPOperation[];
  'wsdl:input': WSDLInput[];
  'wsdl::output': WSDLOutput[];
  'wsdl:fault': WSDLFault[];
}

export interface WSDLServiceAttributes {
  name: string;
}

export interface WSDLService {
  attributes: WSDLServiceAttributes;
  'wsdl:port': WSDLPort[];
}

export interface WSDLPortAttributes {
  binding: string;
  name: string;
}

export interface WSDLPort {
  attributes: WSDLPortAttributes;
  'wsdl:address': WSDLAddress[];
}

export interface WSDLAddressAttributes {
  location: string;
}

export interface WSDLAddress {
  attributes: WSDLAddressAttributes;
}

export interface WSDLImportAttributes {
  location?: string;
  namespace: string;
}

export interface WSDLImport {
  attributes: WSDLImportAttributes;
}

export interface WSDLObject {
  'wsdl:definitions': WSDLDefinition[];
}
