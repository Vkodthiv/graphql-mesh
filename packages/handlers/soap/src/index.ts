import { GetMeshSourceOptions, MeshHandler, YamlConfig, KeyValueCache } from '@graphql-mesh/types';
import { getCachedFetch } from '@graphql-mesh/utils';
import { PredefinedProxyOptions, StoreProxy } from '@graphql-mesh/store';
import { WSDLObject, XSDObject, SOAPLoader } from '@omnigraph/soap';

export default class SoapHandler implements MeshHandler {
  private config: YamlConfig.SoapHandler;
  private cache: KeyValueCache;
  private soapLocationCache: StoreProxy<[string, WSDLObject | XSDObject][]>;

  constructor({ config, baseDir, cache, store, importFn, logger }: GetMeshSourceOptions<YamlConfig.SoapHandler>) {
    this.config = config;
    this.cache = cache;
    this.soapLocationCache = store.proxy('soapLocationCache.json', PredefinedProxyOptions.JsonWithoutValidation);
  }

  async getMeshSource() {
    const soapLocationCacheEntries = await this.soapLocationCache.get();
    const soapLoader = new SOAPLoader({
      fetch: getCachedFetch(this.cache),
    });
    if (soapLocationCacheEntries) {
      for (const [location, object] of soapLocationCacheEntries) {
        soapLoader.loadedLocations.set(location, object);
        if ('schema' in object) {
          for (const schemaObj of object.schema) {
            await soapLoader.loadSchema(schemaObj);
          }
        }
        if ('definitions' in object) {
          for (const definitionObj of object.definitions) {
            await soapLoader.loadDefinition(definitionObj);
          }
        }
      }
    } else {
      await soapLoader.fetchWSDL(this.config.wsdl);
    }
    return {
      schema: soapLoader.buildSchema(),
    };
  }
}
