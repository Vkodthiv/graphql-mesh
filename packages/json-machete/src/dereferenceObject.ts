import { JsonPointer } from 'json-ptr';
import { dirname, isAbsolute, join } from 'path';
import { healJSONSchema } from './healJSONSchema';
import urlJoin from 'url-join';
import { fetch as crossUndiciFetch } from 'cross-undici-fetch';

export const resolvePath = (path: string, root: any): any => {
  return JsonPointer.get(root, path);
};
function isRefObject(obj: any): obj is { $ref: string } {
  return typeof obj === 'object' && obj.$ref;
}

function isURL(str: string) {
  return /^https?:\/\//.test(str);
}

const getAbsolute$Ref = (given$ref: string, baseFilePath: string) => {
  const [givenExternalFileRelativePath, givenRefPath] = given$ref.split('#');
  if (givenExternalFileRelativePath) {
    const cwd = isURL(baseFilePath) ? getCwdForUrl(baseFilePath) : dirname(baseFilePath);
    const givenExternalFilePath = getAbsolutePath(givenExternalFileRelativePath, cwd);
    if (givenRefPath) {
      return `${givenExternalFilePath}#${givenRefPath}`;
    }
    return givenExternalFilePath;
  }
  return `${baseFilePath}#${givenRefPath}`;
};

function getCwdForUrl(url: string) {
  const urlParts = url.split('/');
  urlParts.pop();
  return urlParts.join('/');
}

function normalizeUrl(url: string) {
  return new URL(url).toString();
}

function getAbsolutePath(path: string, cwd: string) {
  if (isURL(path)) {
    return path;
  }
  if (isURL(cwd)) {
    return normalizeUrl(urlJoin(cwd, path));
  }
  if (isAbsolute(path)) {
    return path;
  }
  return join(cwd, path);
}

// eslint-disable-next-line @typescript-eslint/ban-types
export async function dereferenceObject<T extends object, TRoot = T>(
  obj: T,
  {
    cwd = process.cwd(),
    externalFileCache = new Map<string, any>(),
    refMap = new Map<string, any>(),
    root = obj as any,
    fetch = crossUndiciFetch,
  }: {
    cwd?: string;
    externalFileCache?: Map<string, any>;
    refMap?: Map<string, any>;
    root?: TRoot;
    fetch?: WindowOrWorkerGlobalScope['fetch'];
  } = {}
): Promise<T> {
  if (typeof obj === 'object') {
    if (isRefObject(obj)) {
      const $ref = obj.$ref;
      if (refMap.has($ref)) {
        return refMap.get($ref);
      } else {
        const [externalRelativeFilePath, refPath] = $ref.split('#');
        if (externalRelativeFilePath) {
          const externalFilePath = getAbsolutePath(externalRelativeFilePath, cwd);
          const newCwd = isURL(externalFilePath) ? getCwdForUrl(externalFilePath) : dirname(externalFilePath);
          let externalFile = externalFileCache.get(externalFilePath);
          if (!externalFile) {
            const importedJSON = isURL(externalFilePath)
              ? await fetch(externalFilePath).then(res => res.json())
              : await import(externalFilePath);
            externalFile = {};
            for (const key in importedJSON) {
              externalFile[key] = importedJSON[key];
            }
            externalFile = await healJSONSchema(externalFile);
            externalFileCache.set(externalFilePath, externalFile);
          }
          return dereferenceObject(
            refPath
              ? {
                  $ref: `#${refPath}`,
                  ...externalFile,
                }
              : externalFile,
            {
              cwd: newCwd,
              externalFileCache,
              refMap: new Proxy(refMap, {
                get: (originalRefMap, key) => {
                  switch (key) {
                    case 'has':
                      return (given$ref: string) => {
                        const original$Ref = getAbsolute$Ref(given$ref, externalFilePath);
                        return originalRefMap.has(original$Ref);
                      };
                    case 'get':
                      return (given$ref: string) => {
                        const original$Ref = getAbsolute$Ref(given$ref, externalFilePath);
                        return originalRefMap.get(original$Ref);
                      };
                    case 'set':
                      return (given$ref: string, val: any) => {
                        const original$Ref = getAbsolute$Ref(given$ref, externalFilePath);
                        return originalRefMap.set(original$Ref, val);
                      };
                  }
                  throw new Error('Not implemented ' + key.toString());
                },
              }),
              fetch,
            }
          );
        } else {
          const result = resolvePath(refPath, root);
          refMap.set($ref, result);
          return dereferenceObject(result, {
            cwd,
            externalFileCache,
            refMap,
            root,
            fetch,
          });
        }
      }
    } else {
      await Promise.all(
        Object.entries(obj).map(async ([key, val]) => {
          if (typeof val === 'object') {
            obj[key] = await dereferenceObject<any>(val, { cwd, externalFileCache, refMap, root, fetch });
          }
        })
      );
    }
  }
  return obj;
}
