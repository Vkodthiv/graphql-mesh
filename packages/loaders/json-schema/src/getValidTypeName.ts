import { sanitizeNameForGraphQL } from '@graphql-mesh/utils';
import { JSONSchemaObject } from '@json-schema-tools/meta-schema';
import { SchemaComposer } from 'graphql-compose';

export function getValidTypeName({
  schemaComposer,
  isInput,
  subSchema,
}: {
  schemaComposer: SchemaComposer;
  isInput: boolean;
  subSchema: JSONSchemaObject;
}): string {
  const sanitizedName = sanitizeNameForGraphQL(isInput ? subSchema.title + '_Input' : subSchema.title);
  if (schemaComposer.has(sanitizedName)) {
    let i = 2;
    while (schemaComposer.has(sanitizedName + i)) {
      i++;
    }
    return sanitizedName + i;
  }
  return sanitizedName;
}
