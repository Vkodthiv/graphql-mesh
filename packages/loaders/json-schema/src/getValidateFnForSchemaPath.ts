import { memoize2 } from '@graphql-tools/utils';
import { JSONSchema } from 'json-machete';
import Ajv from 'packages/cli/node_modules/ajv/dist/ajv';

const ajvMemoizedCompile = memoize2(function ajvCompile(ajv: Ajv, jsonSchema: JSONSchema) {
  return ajv.compile(jsonSchema);
});

export function getValidateFnForSchemaPath(ajv: Ajv, path: string, schema: JSONSchema) {
  const fakeRefSchema = {
    $ref: `#/definitions/schema${path}`,
    definitions: {
      schema,
    },
  };
  return function validateWithJSONSchema(data: any) {
    const ajvValidate = ajvMemoizedCompile(ajv, fakeRefSchema);
    return ajvValidate(data);
  };
}
