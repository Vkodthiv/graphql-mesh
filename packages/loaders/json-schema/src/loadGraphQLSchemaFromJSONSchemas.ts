import { BaseLoaderOptions } from '@graphql-tools/utils';
import { specifiedDirectives } from 'graphql';
import { dereferenceObject, healJSONSchema, JSONSchemaObject } from 'json-machete';
import { Logger, MeshPubSub } from '@graphql-mesh/types';
import { getComposerFromJSONSchema } from './getComposerFromJSONSchema';
import { SchemaComposer } from 'graphql-compose';
import { DefaultLogger } from '@graphql-mesh/utils';
import { JSONSchemaOperationConfig } from './types';
import { addExecutionLogicToComposer, AddExecutionLogicToComposerOptions } from './addExecutionLogicToComposer';
import { getReferencedJSONSchemaFromOperations } from './getReferencedJSONSchemaFromOperations';

export interface JSONSchemaLoaderOptions extends BaseLoaderOptions {
  baseUrl?: string;
  operationHeaders?: Record<string, string>;
  schemaHeaders?: Record<string, string>;
  operations: JSONSchemaOperationConfig[];
  errorMessage?: string;
  logger?: Logger;
  pubsub?: MeshPubSub;
  fetch?: WindowOrWorkerGlobalScope['fetch'];
  generateInterfaceFromSharedFields?: boolean;
}

export async function getDereferencedJSONSchemaFromOperations({
  operations,
  cwd = process.cwd(),
  logger,
  fetch,
}: {
  operations: JSONSchemaOperationConfig[];
  cwd: string;
  logger: Logger;
  fetch: WindowOrWorkerGlobalScope['fetch'];
}): Promise<JSONSchemaObject> {
  const referencedJSONSchema = await getReferencedJSONSchemaFromOperations({
    operations,
    cwd,
  });
  logger.debug(() => `Dereferencing JSON Schema to resolve all $refs`);
  const fullyDeferencedSchema = await dereferenceObject(referencedJSONSchema, {
    cwd,
    fetch,
  });
  logger.debug(() => `Healing JSON Schema`);
  const healedSchema = await healJSONSchema(fullyDeferencedSchema);
  return healedSchema;
}

export async function loadGraphQLSchemaFromJSONSchemas(name: string, options: JSONSchemaLoaderOptions) {
  const logger = options.logger || new DefaultLogger(name);
  const operations = options.operations;
  const cwd = options.cwd || process.cwd();
  const fullyDeferencedSchema = await getDereferencedJSONSchemaFromOperations({
    operations,
    cwd,
    logger,
    fetch: options.fetch,
  });
  const graphqlSchema = await getGraphQLSchemaFromDereferencedJSONSchema(fullyDeferencedSchema, {
    fetch: options.fetch,
    logger,
    operations,
    operationHeaders: options.operationHeaders,
    baseUrl: options.baseUrl,
    pubsub: options.pubsub,
    errorMessage: options.errorMessage,
    generateInterfaceFromSharedFields: options.generateInterfaceFromSharedFields,
  });
  return graphqlSchema;
}

export async function getGraphQLSchemaFromDereferencedJSONSchema(
  fullyDeferencedSchema: JSONSchemaObject,
  {
    fetch,
    logger,
    operations,
    operationHeaders,
    baseUrl,
    pubsub,
    errorMessage,
    generateInterfaceFromSharedFields,
  }: AddExecutionLogicToComposerOptions & { generateInterfaceFromSharedFields?: boolean }
) {
  logger.debug(() => `Generating GraphQL Schema from the bundled JSON Schema`);
  const visitorResult = await getComposerFromJSONSchema(
    fullyDeferencedSchema,
    logger,
    generateInterfaceFromSharedFields
  );

  const schemaComposerWithoutExecutionLogic = visitorResult.output;

  if (!(schemaComposerWithoutExecutionLogic instanceof SchemaComposer)) {
    throw new Error('The visitor result should be a SchemaComposer instance.');
  }

  // graphql-compose doesn't add @defer and @stream to the schema
  for (const directive of specifiedDirectives) {
    schemaComposerWithoutExecutionLogic.addDirective(directive);
  }

  const schemaComposerWithExecutionLogic = await addExecutionLogicToComposer(schemaComposerWithoutExecutionLogic, {
    fetch,
    logger,
    operations,
    operationHeaders,
    baseUrl,
    pubsub,
    errorMessage,
  });

  return schemaComposerWithExecutionLogic.buildSchema();
}
