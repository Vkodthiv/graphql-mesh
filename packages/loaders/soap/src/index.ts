import { sanitizeNameForGraphQL } from '@graphql-mesh/utils';
import { parse as parseXML, X2jOptions } from 'fast-xml-parser';
import { InputTypeComposer, InputTypeComposerFieldConfigMapDefinition, SchemaComposer } from 'graphql-compose';
import { XSComplexType, WSDLDefinition, WSDLObject, WSDLPortType, WSDLBinding, WSDLMessage, XSSchema } from './types';

export interface SOAPLoaderOptions {
  fetch: WindowOrWorkerGlobalScope['fetch'];
}

const PARSE_XML_OPTIONS: Partial<X2jOptions> = {
  attributeNamePrefix: '',
  attrNodeName: 'attributes',
  textNodeName: 'innerText',
  ignoreAttributes: false,
  ignoreNameSpace: false,
  arrayMode: true,
  allowBooleanAttributes: true,
};

export class SOAPLoader {
  private schemaComposer = new SchemaComposer();
  private namespaceDefinitionsMap = new Map<string, WSDLDefinition[]>();
  private namespaceComplexTypesMap = new Map<string, Map<string, XSComplexType>>();
  private namespacePortTypesMap = new Map<string, Map<string, WSDLPortType>>();
  private namespaceBindingMap = new Map<string, Map<string, WSDLBinding>>();
  private namespaceMessageMap = new Map<string, Map<string, WSDLMessage>>();
  private aliasMap = new WeakMap<any, Map<string, string>>();
  private messageInputTCMap = new WeakMap<WSDLMessage, InputTypeComposer>();
  private complexTypeInputTCMap = new WeakMap<XSComplexType, InputTypeComposer>();
  private namespaceTypePrefixMap = new Map<string, string>();

  constructor(private options: SOAPLoaderOptions) {}
  private getNamespaceDefinitions(namespace: string) {
    let namespaceDefinitions = this.namespaceDefinitionsMap.get(namespace);
    if (!namespaceDefinitions) {
      namespaceDefinitions = [];
      this.namespaceDefinitionsMap.set(namespace, namespaceDefinitions);
    }
    return namespaceDefinitions;
  }

  private getNamespaceComplexTypeMap(namespace: string) {
    let namespaceComplexTypes = this.namespaceComplexTypesMap.get(namespace);
    if (!namespaceComplexTypes) {
      namespaceComplexTypes = new Map();
      this.namespaceComplexTypesMap.set(namespace, namespaceComplexTypes);
    }
    return namespaceComplexTypes;
  }

  private getNamespacePortTypeMap(namespace: string) {
    let namespacePortTypes = this.namespacePortTypesMap.get(namespace);
    if (!namespacePortTypes) {
      namespacePortTypes = new Map();
      this.namespacePortTypesMap.set(namespace, namespacePortTypes);
    }
    return namespacePortTypes;
  }

  private getNamespaceBindingMap(namespace: string) {
    let namespaceBindingMap = this.namespaceBindingMap.get(namespace);
    if (!namespaceBindingMap) {
      namespaceBindingMap = new Map();
      this.namespaceBindingMap.set(namespace, namespaceBindingMap);
    }
    return namespaceBindingMap;
  }

  private getNamespaceMessageMap(namespace: string) {
    let namespaceMessageMap = this.namespaceMessageMap.get(namespace);
    if (!namespaceMessageMap) {
      namespaceMessageMap = new Map();
      this.namespaceMessageMap.set(namespace, namespaceMessageMap);
    }
    return namespaceMessageMap;
  }

  async loadWSDL(location: string) {
    const response = await this.options.fetch(location);
    const wsdlText = await response.text();
    const wsdlObject: WSDLObject = parseXML(wsdlText, PARSE_XML_OPTIONS);
    for (const definition of wsdlObject['wsdl:definitions']) {
      this.getNamespaceDefinitions(definition.attributes.targetNamespace).push(definition);
      this.namespaceTypePrefixMap.set(definition.attributes.targetNamespace, definition.attributes.name);
      if (definition['wsdl:import']) {
        for (const importObj of definition['wsdl:import']) {
          if (importObj.attributes.location) {
            await this.loadWSDL(importObj.attributes.location);
          }
        }
      }
      if (definition['wsdl:types']) {
        for (const typesObj of definition['wsdl:types']) {
          for (const schemaObj of typesObj['xs:schema']) {
            const namespace = schemaObj.attributes.targetNamespace;
            const aliasMap = this.getAliasMapFromAttributes(schemaObj.attributes);
            const namespaceComplexTypes = this.getNamespaceComplexTypeMap(namespace);
            for (const complexType of schemaObj['xs:complexType']) {
              namespaceComplexTypes.set(complexType.attributes.name, complexType);
              this.aliasMap.set(complexType, aliasMap);
            }
          }
        }
      }
      const definitionAliasMap = this.getAliasMapFromAttributes(definition.attributes);
      if (definition['wsdl:portType']) {
        const namespacePortTypes = this.getNamespacePortTypeMap(definition.attributes.targetNamespace);
        for (const portTypeObj of definition['wsdl:portType']) {
          namespacePortTypes.set(portTypeObj.attributes.name, portTypeObj);
          this.aliasMap.set(portTypeObj, definitionAliasMap);
        }
      }
      if (definition['wsdl:binding']) {
        const namespaceBindingMap = this.getNamespaceBindingMap(definition.attributes.targetNamespace);
        for (const bindingObj of definition['wsdl:binding']) {
          namespaceBindingMap.set(bindingObj.attributes.name, bindingObj);
          this.aliasMap.set(bindingObj, definitionAliasMap);
        }
      }
      if (definition['wsdl:message']) {
        const namespaceMessageMap = this.getNamespaceMessageMap(definition.attributes.targetNamespace);
        for (const messageObj of definition['wsdl:message']) {
          namespaceMessageMap.set(messageObj.attributes.name, messageObj);
          this.aliasMap.set(messageObj, definitionAliasMap);
        }
      }
    }
  }

  getAliasMapFromAttributes(attributes: XSSchema['attributes'] | WSDLDefinition['attributes']) {
    const aliasMap = new Map<string, string>();
    for (const attributeName in attributes) {
      if (attributeName.startsWith('xmlns:')) {
        const alias = attributeName.split('xmlns:')[1];
        aliasMap.set(alias, attributes[attributeName]);
      }
    }
    return aliasMap;
  }

  getInputTypeForTypeNameInNamespace({ typeName, typeNamespace }: { typeName: string; typeNamespace: string }) {
    const complexType = this.getNamespaceComplexTypeMap(typeNamespace)?.get(typeName);
    if (complexType) {
      return this.getInputTypeForComplexType(complexType, typeNamespace);
    }
    throw new Error(`Type: ${typeName} couldn't be found in ${typeNamespace}`);
  }

  getInputTypeForComplexType(complexType: XSComplexType, complexTypeNamespace: string) {
    let complexTypeTC = this.complexTypeInputTCMap.get(complexType);
    if (!complexTypeTC) {
      const complexTypeName = complexType.attributes.name;
      const prefix = this.namespaceTypePrefixMap.get(complexTypeNamespace);
      complexTypeTC = this.schemaComposer.createInputTC({
        name: `${prefix}_${complexTypeName}_Input`,
        fields: () => {
          const aliasMap = this.aliasMap.get(complexType);
          const fieldMap: InputTypeComposerFieldConfigMapDefinition = {};
          for (const sequenceObj of complexType['xs:sequence']) {
            for (const elementObj of sequenceObj['xs:element']) {
              const [typeNamespaceAlias, typeName] = elementObj.attributes.type.split(':');
              const typeNamespace = aliasMap.get(typeNamespaceAlias);
              if (!typeNamespace) {
                throw new Error(`Namespace alias: ${typeNamespace} is undefined!`);
              }
              fieldMap[elementObj.attributes.name] = {
                type: this.getInputTypeForTypeNameInNamespace({ typeName, typeNamespace }),
              };
            }
          }
          for (const complexContentObj of complexType['xs:complexContent']) {
            for (const extensionObj of complexContentObj['xs:extension']) {
              const [baseTypeNamespaceAlias, baseTypeName] = extensionObj.attributes.base.split(':');
              const baseTypeNamespace = aliasMap.get(baseTypeNamespaceAlias);
              if (!baseTypeNamespace) {
                throw new Error(`Namespace alias: ${baseTypeNamespace} is undefined!`);
              }
              const baseTypeTC = this.getInputTypeForTypeNameInNamespace({
                typeName: baseTypeName,
                typeNamespace: baseTypeNamespace,
              });
              for (const fieldName in baseTypeTC.getFields()) {
                fieldMap[fieldName] = baseTypeTC.getField(fieldName);
              }
              for (const sequenceObj of extensionObj['xs:sequence']) {
                for (const elementObj of sequenceObj['xs:element']) {
                  const [typeNamespaceAlias, typeName] = elementObj.attributes.type.split(':');
                  const typeNamespace = aliasMap.get(typeNamespaceAlias);
                  if (!typeNamespace) {
                    throw new Error(`Namespace alias: ${typeNamespace} is undefined!`);
                  }
                  fieldMap[elementObj.attributes.name] = {
                    type: this.getInputTypeForTypeNameInNamespace({ typeName, typeNamespace }),
                  };
                }
              }
            }
          }
          return fieldMap;
        },
      });
    }
    return complexTypeTC;
  }

  getInputTypeForMessage(message: WSDLMessage, messageNamespace: string) {
    let inputTC = this.messageInputTCMap.get(message);
    if (!inputTC) {
      const messageName = message.attributes.name;
      const prefix = this.namespaceTypePrefixMap.get(messageNamespace);
      const inputTCName = sanitizeNameForGraphQL(`${prefix}_${messageName}_Input`);
      inputTC = this.schemaComposer.createInputTC({
        name: inputTCName,
        fields: () => {
          const fieldMap: InputTypeComposerFieldConfigMapDefinition = {};
          const aliasMap = this.aliasMap.get(message);
          for (const part of message['wsdl:part']) {
            const partName = part.attributes.name;
            const [typeNamespaceAlias, typeName] = part.attributes.element.split(':');
            const typeNamespace = aliasMap.get(typeNamespaceAlias);
            if (!typeNamespace) {
              throw new Error(`Namespace alias: ${typeNamespace} is undefined!`);
            }
            fieldMap[partName] = {
              type: this.getInputTypeForTypeNameInNamespace({ typeName, typeNamespace }),
            };
          }
          return fieldMap;
        },
      });
      this.messageInputTCMap.set(message, inputTC);
    }
    return inputTC;
  }

  build() {
    for (const [_namespace, definitions] of this.namespaceDefinitionsMap) {
      // const prefix = this.namespaceTypePrefixMap.get(namespace);
      for (const definition of definitions) {
        const serviceAndPortAliasMap = this.getAliasMapFromAttributes(definition.attributes);
        if (definition['wsdl:service']) {
          for (const serviceObj of definition['wsdl:service']) {
            const serviceName = serviceObj.attributes.name;
            for (const portObj of serviceObj['wsdl:port']) {
              const portName = portObj.attributes.name;
              const [bindingNamespaceAlias, bindingName] = portObj.attributes.binding.split(':');
              const bindingNamespace = serviceAndPortAliasMap.get(bindingNamespaceAlias);
              if (!bindingNamespace) {
                throw new Error(`Namespace alias: ${bindingNamespaceAlias} is undefined!`);
              }
              const bindingObj = this.getNamespaceBindingMap(bindingNamespace).get(bindingName);
              if (!bindingObj) {
                throw new Error(
                  `Binding: ${bindingName} is not defined in ${bindingNamespace} needed for ${serviceName}->${portName}`
                );
              }
              const bindingAliasMap = this.aliasMap.get(bindingObj);
              if (!bindingAliasMap) {
                throw new Error(`Namespace alias definitions couldn't be found for ${bindingName}`);
              }
              const [portTypeNamespaceAlias, portTypeName] = bindingObj.attributes.type.split(':');
              const portTypeNamespace = bindingAliasMap.get(portTypeNamespaceAlias);
              if (!portTypeNamespace) {
                throw new Error(`Namespace alias: ${portTypeNamespace} is undefined!`);
              }
              const portTypeObj = this.getNamespacePortTypeMap(portTypeNamespace).get(portTypeName);
              if (!portTypeObj) {
                throw new Error(
                  `Port Type: ${portTypeName} is not defined in ${portTypeNamespace} needed for ${bindingNamespaceAlias}->${bindingName}`
                );
              }
              const portTypeAliasMap = this.aliasMap.get(portTypeObj);
              for (const operationObj of portTypeObj['wsdl:operation']) {
                const operationName = operationObj.attributes.name;
                const inputTypes = new Set<InputTypeComposer>();
                for (const inputObj of operationObj['wsdl:input']) {
                  const [messageNamespaceAlias, messageName] = inputObj.attributes.message.split(':');
                  const messageNamespace = portTypeAliasMap.get(messageNamespaceAlias);
                  if (!messageNamespace) {
                    throw new Error(`Namespace alias: ${messageNamespace} is undefined!`);
                  }
                  const messageObj = this.getNamespaceMessageMap(messageNamespace).get(messageName);
                  if (!messageObj) {
                    throw new Error(
                      `Message: ${messageName} is not defined in ${messageNamespace} needed for ${portTypeName}->${operationName}`
                    );
                  }
                  const messageInputTC = this.getInputTypeForMessage(messageObj, messageNamespace);
                  inputTypes.add(messageInputTC);
                }
              }
            }
          }
        }
      }
    }
  }
}
