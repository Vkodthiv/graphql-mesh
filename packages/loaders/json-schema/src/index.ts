import { loadGraphQLSchemaFromJSONSchemas } from './loadGraphQLSchemaFromJSONSchemas';
export default loadGraphQLSchemaFromJSONSchemas;
export * from './addExecutionLogicToComposer';
export * from './getReferencedJSONSchemaFromOperations';
export * from './getComposerFromJSONSchema';
export * from './loadGraphQLSchemaFromJSONSchemas';
export * from './types';
