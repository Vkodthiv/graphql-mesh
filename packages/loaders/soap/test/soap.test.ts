import { SOAPLoader } from '../src';
import { fetch } from 'cross-undici-fetch';
import { execute, parse, printSchema } from 'graphql';

describe('SOAP Loader', () => {
  it('should generate an executable schema', async () => {
    const soapLoader = new SOAPLoader({
      fetch: fetch as any,
    });
    soapLoader.loadXMLSchemaNamespace();
    await soapLoader.loadWSDL('https://www.w3schools.com/xml/tempconvert.asmx?WSDL');
    soapLoader.addRootFieldsToComposer();
    const schema = soapLoader.buildSchema();
    expect(printSchema(schema)).toMatchSnapshot();
    expect(
      await execute({
        schema,
        document: parse(`
          mutation {
            tns_TempConvert_TempConvertSoap_CelsiusToFahrenheit(CelsiusToFahrenheit: { Celsius: "24" }) {
              CelsiusToFahrenheitResponse {
                CelsiusToFahrenheitResult
              }
            }
          }
        `),
      })
    ).toMatchInlineSnapshot(`
      Object {
        "data": Object {
          "tns_TempConvert_TempConvertSoap_CelsiusToFahrenheit": Object {
            "CelsiusToFahrenheitResponse": Object {
              "CelsiusToFahrenheitResult": "75.2",
            },
          },
        },
      }
    `);
  });
});
