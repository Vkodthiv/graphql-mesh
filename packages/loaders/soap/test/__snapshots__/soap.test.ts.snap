// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SOAP Loader should generate an executable schema 1`] = `
"type Query {
  placeholder: Void
}

\\"\\"\\"Represents NULL values\\"\\"\\"
scalar Void

type Mutation {
  tns_TempConvert_TempConvertSoap_FahrenheitToCelsius(FahrenheitToCelsius: tns_FahrenheitToCelsius_Input): tns_FahrenheitToCelsiusSoapOut
  tns_TempConvert_TempConvertSoap_CelsiusToFahrenheit(CelsiusToFahrenheit: tns_CelsiusToFahrenheit_Input): tns_CelsiusToFahrenheitSoapOut
  tns_TempConvert_TempConvertSoap12_FahrenheitToCelsius(FahrenheitToCelsius: tns_FahrenheitToCelsius_Input): tns_FahrenheitToCelsiusSoapOut
  tns_TempConvert_TempConvertSoap12_CelsiusToFahrenheit(CelsiusToFahrenheit: tns_CelsiusToFahrenheit_Input): tns_CelsiusToFahrenheitSoapOut
  tns_TempConvert_TempConvertHttpPost_FahrenheitToCelsius(Fahrenheit: String): tns_FahrenheitToCelsiusHttpPostOut
  tns_TempConvert_TempConvertHttpPost_CelsiusToFahrenheit(Celsius: String): tns_CelsiusToFahrenheitHttpPostOut
}

type tns_FahrenheitToCelsiusSoapOut {
  FahrenheitToCelsiusResponse: tns_FahrenheitToCelsiusResponse
}

type tns_FahrenheitToCelsiusResponse {
  FahrenheitToCelsiusResult: String
}

input tns_FahrenheitToCelsius_Input {
  Fahrenheit: String
}

type tns_CelsiusToFahrenheitSoapOut {
  CelsiusToFahrenheitResponse: tns_CelsiusToFahrenheitResponse
}

type tns_CelsiusToFahrenheitResponse {
  CelsiusToFahrenheitResult: String
}

input tns_CelsiusToFahrenheit_Input {
  Celsius: String
}

type tns_FahrenheitToCelsiusHttpPostOut {
  string: String
}

type tns_CelsiusToFahrenheitHttpPostOut {
  string: String
}"
`;
